prdoc/pr_3331.prdoc

```diff
@@ -0,0 +1,13 @@
+# Schema: Polkadot SDK PRDoc Schema (prdoc) v1.0.0
+# See doc at https://raw.githubusercontent.com/paritytech/polkadot-sdk/master/prdoc/schema_user.json
+
+title: Use Relay Blocknumber in Pallet Broker
+
+doc:
+  - audience: Runtime Dev
+    description: |
+      Changing `sale_start`, `interlude_length` and `leading_length` in `pallet_broker` to use relay chain block numbers instead of parachain block numbers.
+      Relay chain block numbers are almost deterministic and more future proof.
```

Comment:
I would add a note to warn people who are already running with a configured broker pallet that with this change they will potentially have different real-world leadin and interlude times. If they previously were timing these against 12s blocks on their parachain they will now be timed by 6s blocks, and therefore the interlude and leadin period will pass twice as quickly. These can be updated using the `configure` extrinsic on the broker pallet.

There is an argument to be made for a migration here to automatically fix the timings, but I don't have a strong opinion on it. Maybe @bkchr or @ggwpez would

Reply:
Actually, with this change we will definitely need a migration, as they are in general totally different types.
All the rest of the changes are looking great

Reply:
AFAIK, the `BlocknumberFor<T>` and `RelayBlockNumberFor<T>` are both configured as `u32` ubiquitously in the ecosystem, so I don't think any migration is required.
Am I missing something?

Reply:
FRAME is a toolkit which is designed to be as generic as possible. Making changes in a pallet can make no assumptions about how it is configured except for those implied by any trait bounds and so in general these are different types and should be treated as such for any changes to the pallet.

Reply:
Okay, make sense.

So the migration will be from the type `BlockNumberFor<T>` to  `RelayBlockNumberOf<T>` for storage variables 
`Configuration<T>` and `SaleInfo<T>`.

Both these blocknumber types implements trait `Atleast32BitUnsigned`.

My question is how are we supposed to handle cases where `BlockNumberFor<T>` is a larger type than `RelayBlockNumberOf<T>`?
For example converting u128 to u32 where the current `BlockNumberFor<T>` > u32::MAX

We can use `TryFrom` but how to handle the case where it fails?

Reply:
Remember that these configurations are defining the length of sections of a sale, if the types are `Atleast32BitUnsigned`, then values that are too big will be defining sales that are at least `4_294_967_295` blocks long, and we can assume this is a misconfiguration, as that would be a pretty long sale! I think in this case we can log it as an error and insert some default value (maybe u32::MAX, as zero is not meaningful). The config would then be fixed through governance. A similar argument goes for the sale start.